src/WebJobs.Extensions.CosmosDB/Trigger/CosmosDBMetricsProvider.cs

```diff
@@ -0,0 +1,135 @@
+                {
+                    while (iterator.HasMoreResults)
+                    {
+                        FeedResponse<ChangeFeedProcessorState> response = await iterator.ReadNextAsync();
+                        partitionWorkList.AddRange(response);
+                    }
+                }
+
+                partitionCount = partitionWorkList.Count;
+                remainingWork = partitionWorkList.Sum(item => item.EstimatedLag);
+            }
+            catch (Exception e) when (e is CosmosException || e is InvalidOperationException)
+            {
+                if (!TryHandleCosmosException(e))
+                {
+                    _logger.LogWarning(Events.OnScaling, "Unable to handle {0}: {1}", e.GetType().ToString(), e.Message);
+                    if (e is InvalidOperationException)
+                    {
+                        throw;
+                    }
+                }
+            }
+            catch (System.Net.Http.HttpRequestException e)
+            {
+                string errormsg;
+
+                var webException = e.InnerException as WebException;
+                if (webException != null &&
+                    webException.Status == WebExceptionStatus.ProtocolError)
+                {
+                    string statusCode = ((HttpWebResponse)webException.Response).StatusCode.ToString();
+                    string statusDesc = ((HttpWebResponse)webException.Response).StatusDescription;
+                    errormsg = string.Format("CosmosDBTrigger status {0}: {1}.", statusCode, statusDesc);
+                }
+                else if (webException != null &&
+                    webException.Status == WebExceptionStatus.NameResolutionFailure)
+                {
+                    errormsg = string.Format("CosmosDBTrigger Exception message: {0}.", webException.Message);
+                }
+                else
+                {
+                    errormsg = e.ToString();
+                }
+
+                _logger.LogWarning(Events.OnScaling, errormsg);
+            }
+
+            return new CosmosDBTriggerMetrics
+            {
+                Timestamp = DateTime.UtcNow,
+                PartitionCount = partitionCount,
+                RemainingWork = remainingWork
+            };
+        }
+
+        // Since all exceptions in the Cosmos client are thrown as CosmosExceptions, we have to parse their error strings because we dont have access to the internal types
+        private bool TryHandleCosmosException(Exception exception)
+        {
+            string errormsg = null;
+            string exceptionMessage = exception.Message;
+
+            if (!string.IsNullOrEmpty(exceptionMessage))
+            {
+                foreach (KeyValuePair<string, string> exceptionString in KnownDocumentClientErrors)
+                {
+                    if (exceptionMessage.IndexOf(exceptionString.Key, StringComparison.OrdinalIgnoreCase) >= 0)
```

Comment:
Thanks for the callout! It might be better to separate this part into a different PR as I do not have context on the code here (was copy/paste from existing scale monitor), just to keep changes separate from TBS

Reply:
Completely agree, I was just writing it for our future selves